### PR TITLE
Move development tool configuration to `pyproject.toml`

### PR DESCRIPTION
**Pull request by brocksam**
#### References to other Issues or PRs

Partially addresses #24762.

#### Brief description of what is fixed or changed
This PR starts the migration towards using `pyproject.toml` and PEP 621 compliance.

The contents of `pytest.ini` and `setup.cfg` are moved and these two files deleted. Where possible (pytest configuration that was in `pytest.ini`, and mypy configuration that was in `setup.cfg`) has been moved to `pyproject.toml`. The contents should remain identical with only the syntax being updated to TOML. 

The project metadata that was in `setup.cfg` (long description only) has been moved back to `setup.py` for the time being. This is because I was seeing some failures in CI (specifically the Pyodide job) when I tried to split the metadata between `setup.py` and `pyproject.toml`, so I think it's best left to translate the whole of `setup.py` over to `pyproject.toml` in a separate PR.

[flake8 does not support configuration in `pyproject.toml`](https://github.com/PyCQA/flake8/issues/234). This PR moved the flake8 configuration from `setup.cfg` to a new `.flake8` file. The reason for this is that it allows the `setup.cfg` file to be removed which should help promote using `pyproject.toml` as the default going forward. A `.flake8` file was the preferable alternative as only flake8 configuration can be put in here, and if/when flake8 is removed in favour of Ruff (#24508) then this will simply just require that `.flake8` be deleted.

#### Other comments
A further PR will follow later to address the remainder of #24762 and `setup.py`.

#### Release Notes

NO ENTRY

### CODE REVIEW

**Review by Skylion007 — APPROVED** [review]

**Comment by oscarbenjamin** [comment]
I'd be interested to know how these changes fair on the 1.12 branch so that the GitHub Actions release workflow runs on them. I'm currently fixing the setup on the 1.12 branch though.

Could you open a separate PR against the 1.12 branch (rather than master)?

Otherwise these changes look good to me.

**Comment by oscarbenjamin** [comment]
> Could you open a separate PR against the 1.12 branch (rather than master)?

To be clear I mean keep this PR but also open another one. I would do it by creating a new branch from 1.12 and then cherry picking the commits from this PR:
```
git fetch upstream
git checkout 1.12
git checkout -b pr_branch
git cherry-pick <commit>...
```
For `<commit>` the simple thing is just to run cherry-pick once for each commit. Otherwise you can use a commit range to specify the commits that are in this PR (as opposed to those from the base branch).

File: pyproject.toml [hunk 1]
```diff
@@ -53,3 +76,53 @@ dummy-variable-rgx = "^(_+|(_+[a-zA-Z0-9_]*[a-zA-Z0-9]+?))$"
...
+    "sympy.benchmarks.*",
+    "sympy.plotting.pygletplot.*",
+]
+ignore_errors = true
+
+# Third-party untyped code:
+[[tool.mypy.overrides]]
+module = [
+    "antlr4.*",
+    "Cython.*",
```
Comment (oscarbenjamin) [inline pyproject.toml:101]: Side note: looks like some of these can be removed:

numpy, PIL, python-sat, pytest, _pytest, aesara, xml, pydy

Many of the others seem to have chosen not to expose py.typed files which I now regret agreeing to for sympy (gh-22337).

**Comment by oscarbenjamin** [comment]
Looks good.

**Pull request merged**